### DISCUSSION TITLE
Jar package dependency scanning is misleading

### DISCUSSION THREAD

**Discussion by zangcc** [aquasecurity/trivy#7330] [Q&A] [answered]
### Description

Take `com.google.protobuf:protobuf-java:3.16.1` as an example. There is no `com.google.protobuf:protobuf-java:3.16.1 `version in my code project, but Trivy's scan results show `com.google.protobuf:protobuf-java:3.16.1`, which makes us misunderstand that this is a false positive.

The reason why `com.google.protobuf:protobuf-java:3.16.1 `appears in Trivy's scan results is because `com.google.protobuf:protobuf-java:3.16.1 `is built-in by `web.jar/BOOT-INF/lib/nacos-client-2.0.4.jar.`This is what I scanned with the following command: `./trivy rootfs -f json --list-all-pkgs web.jar -o result.txt`
![image](https://github.com/user-attachments/assets/40969da8-81ec-457c-aa36-d081b27ea36e)

I think Trivy should optimize this aspect. Because the scan result is not very realistic, I can't find the corresponding version of the vulnerability in the `mvn dependency tree`.

### Desired Behavior

I hope Trivy can automatically declare where the jar package comes from in the scan results of the jar package type.

### Actual Behavior

Trivy can accurately print out the version of the jar package, but this can be easily misunderstood. For example, although I did not declare this low-version jar package in my project, this low-version jar package actually exists in the scan results. I need to use the command` --list-all-pkgs `for secondary confirmation, which is very troublesome.

### Reproduction Steps

```bash
./trivy rootfs -f json --list-all-pkgs web.jar -o result.txt
```

### Target

None

### Scanner

None

### Output Format

None

### Mode

None

### Debug Output

```bash
none
```

### Operating System

macOS

### Version

```bash
trivy_0.48.1_macOS-64bit
```

### Checklist

- [ ] Run `trivy clean --all`
- [ ] Read [the troubleshooting](https://aquasecurity.github.io/trivy/latest/docs/references/troubleshooting/)

**Comment by DmitriyLewen** [comment]
Hello @zangcc 
Thanks for your report!

Your `web.jar` contains `nacos-client-2.0.4.jar` file.
`nacos-client-2.0.4.jar` contains `pom.properties` file with info about `com.google.protobuf:protobuf-java:3.16.1`:
```bash
➜ cat META-INF/maven/com.google.protobuf/protobuf-java/pom.properties 
#Generated by org.apache.felix.bundleplugin
#Wed Jan 05 22:11:00 UTC 2022
version=3.16.1
groupId=com.google.protobuf
artifactId=protobuf-java
```
This means that source code of `nacos-client-2.0.4.jar` contains `com.google.protobuf:protobuf-java:3.16.1`.

That is why Trivy show this package and vulnerability for this package.

Regards, Dmitriy

**Reply by zangcc** [reply]
Thank you for your reply. Yes, as you said, I believe Trivy's scan results are accurate. 

So besides the `--list-all-pkgs` parameter that can list the dependency transitive relationships of the jar, is there any better way to list them in the scan results? (Similar to the `--dependency-tree` of Maven project scanning, the scan results are very clear)

After all, scanning with the `--list-all-pkgs` parameter can only output json type reports, and json type reports do not seem as convenient and concise as txt type output reports scanned with the `--dependency-tree` parameter.

**Reply by DmitriyLewen** [reply, marked as answer]
Unfortunately, we don't support `-dependency-tree` flag for `jar` files.
You can read about this in https://github.com/aquasecurity/trivy/issues/5486

So `--list-all-pkgs` is better way to find location of dependency.

**Reply by zangcc** [reply]
Thanks for your reply. Will Trivy consider optimizing this part in the future? Because there are too many results scanned with the `--list-all-pkgs` parameter, I actually only want to care about the dependency transitive relationship in the jar package, and I want to highlight this point in the output report.

**Reply by DmitriyLewen** [reply]
> I actually only want to care about the dependency transitive relationship in the jar package

Do you mean including only indirect dependencies in the report?

> Will Trivy consider optimizing this part in the future? Because there are too many results scanned with the --list-all-pkgs parameter

I think not. We try not to overload it with unnecessary information, so that it is compact and easy to read.
for additional information there is a json format

**Reply by zangcc** [reply]
> Do you mean including only indirect dependencies in the report?

Yes, for example: `--list-all-pkgs-only`. (I'm just making a suggestion)

**Reply by DmitriyLewen** [reply]
we added `--pkg-relationship` flag. (https://github.com/aquasecurity/trivy/pull/7237) (`v0.55.0` will include this flag)
Maybe this is what you need.

**Reply by knqyf263** [reply]
v0.54.0 includes it.
https://github.com/aquasecurity/trivy/pull/7075

**Reply by DmitriyLewen** [reply]
oops.. that was my mistake.
@knqyf263 Thanks for correcting me.
